### ISSUE TITLE
[Bug]: ScanImageImagingInterface fails to read any ScanImage2021 metadata

### ISSUE THREAD

**Issue by simon-ball** [catalystneuro/neuroconv#385]
### What happened?

Attempting to read metadata from a ScanImage generated Tiff fails to extract metadata such as the scanner sampling frequency. Stepping through [the code](https://github.com/catalystneuro/neuroconv/blob/main/src/neuroconv/datainterfaces/ophys/scanimage/scanimageimaginginterface.py#L11) for `neuroconv.datainterfaces.ophys.scanimage.scanimageimaginginterface.extract_extra_metadata` indicates that it can't, in fact, read _any_ of the `SI` flagged metadata tags that are written.

As far as I can see, the root cause is as follows:

`extract_extra_metadata` looks like so:
```python
description = ScanImageTiffReader.ScanImageTiffReader(str(file_path)).description(iframe=0)
extra_metadata = {x.split("=")[0]: x.split("=")[1] for x in description.split("\r") if "=" in x}
```

We can step through that with an example file (that I can provide, but it's both massive, and I would rather provide it direct to you than publicly)
```python
description
>>> 'frameNumbers = 1\nacquisitionNumbers = 1\nframeNumberAcquisition = 1\nframeTimestamps_sec = 0.000000000\nacqTriggerTimestamps_sec = -0.000088640\nnextFileMarkerTimestamps_sec = -1.000000000\nendOfAcquisition = 0\nendOfAcquisitionMode = 0\ndcOverVoltage = 0\nepoch = [2023  2  6 10 51 13.374]\nauxTrigger0 = []\nauxTrigger1 = []\nauxTrigger2 = []\nauxTrigger3 = []\nI2CData = {}\n'
```

The metadata extraction fails at the first hurdle, because the string isn't delimited by carriage return, but by newline (this might vary depending on the operating system ScanImage is running on, maybe?). So nearly all of the metadata is dropped, and what isn't dropped, is corrupted. 
```python
extra_metadata
>>> {'frameNumbers ': ' 1\nacquisitionNumbers '}
```

No metadata matching the pattern that the [ScanImageExtractorInterface](https://github.com/catalystneuro/neuroconv/blob/main/src/neuroconv/datainterfaces/ophys/scanimage/scanimageimaginginterface.py#L51) searches for  is present:
```python
        if "state.acq.frameRate" in self.image_metadata:
            sampling_frequency = float(self.image_metadata["state.acq.frameRate"])
```

Notably, none of the SI metadata (including the framerate) is available via the Interface, because it's not under `ScanImageTiffReader().description`, but under `ScanImageTiffReader().metadata`

```python
tr = ScanImageTiffReader.ScanImageTiffReader(path.as_posix())
description = tr.description(iframe=0)
extra_metadata = {x.split("=")[0]: x.split("=")[1] for x in description.split("\r") if "=" in x}
tr.metadata().split("\n")
>>> ['SI.LINE_FORMAT_VERSION = 1',
 'SI.PREMIUM = false',
 'SI.TIFF_FORMAT_VERSION = 4',
 "SI.VERSION_COMMIT = '3baa4b74bec9938bfd4c7403bddbe78f2a0ccd93'",
 'SI.VERSION_MAJOR = 2021',
 'SI.VERSION_MINOR = 0',
 'SI.VERSION_UPDATE = 0',
<...>
'SI.hRoiManager.scanFrameRate = 14.4926',
'SI.hRoiManager.scanVolumeRate = 7.24632',
<...>
'SI.hScan2D.scannerFrequency = 1999.98',
<...>
]
```
(about 380 lines of metadata omitted for brevity, but the scanner sample frequency data is present)

===

_To the best of my knowledge_, the researchers recording this data are not using any special sauce, it's just an out-of-the-box ScanImage2021 installation (see the SI.VERSION.* tags in the metadata above). As far as I am aware, we have used approximately the same code for extracting SI metadata from tifs generated from as far back as ScanImage2016. 

We typically use a fork of the [atlab/scanreader](https://github.com/atlab/scanreader) package to extract this metadata (most of the forked changes are about keeping up to date with new SI versions), but all of the tags can be read out via ScanImageTiffReader too. 

To the best of my knowledge, we invariably run ScanImage on Windows 10 workstations. Neuroconv is running on an x86 Ubuntu VM, since the DeepLabCut dependency appears to be incompatible with an M1 architecture.

### Steps to Reproduce

```python
from neuroconv.datainterfaces import ScanImageImagingInterface
from pathlib import Path
my_data_diectory = Path("/network/share/subject/date")
file_paths = my_data_directory.glob("*.tif")
interface_si = [ScanImageImagingInterface(file_path=fp) for fp in file_paths]

>>> AssertionError: sampling frequency not found in image metadata, input the frequency using the argument `fallback_sampling_frequency`
```

### Traceback

```python
File ~/miniconda3/envs/exporter/lib/python3.10/site-packages/neuroconv/datainterfaces/ophys/scanimage/scanimageimaginginterface.py:57, in ScanImageImagingInterface.__init__(self, file_path, fallback_sampling_frequency, verbose)
     52 else:
     53     assert_msg = (
     54         "sampling frequency not found in image metadata, "
     55         "input the frequency using the argument `fallback_sampling_frequency`"
     56     )
---> 57     assert fallback_sampling_frequency is not None, assert_msg
     58     sampling_frequency = fallback_sampling_frequency
     60 super().__init__(file_path=file_path, sampling_frequency=sampling_frequency, verbose=verbose)

AssertionError: sampling frequency not found in image metadata, input the frequency using the argument `fallback_sampling_frequency`
```

### Operating System

Linux

### Python Executable

Python

### Python Version

3.8

### Package Versions

absl-py==1.4.0
appdirs==1.4.4
arrow==1.2.3
asciitree==0.3.3
astropy==5.2.1
asttokens==2.2.1
astunparse==1.6.3
attrs==22.2.0
backcall==0.2.0
beautifulsoup4==4.12.0
bidsschematools==0.6.0
blessings==1.7
blosc2==2.0.0
boto3==1.24.96
botocore==1.27.96
cachetools==5.3.0
certifi @ file:///croot/certifi_1671487769961/work/certifi
cffi==1.15.1
charset-normalizer==3.1.0
ci-info==0.3.0
click==8.1.3
click-didyoumean==0.3.0
comm==0.1.3
contourpy==1.0.7
cryptography==40.0.1
cycler==0.11.0
Cython==0.29.33
dandi==0.51.0
dandischema==0.8.2
datajoint==0.14.0
debugpy==1.6.6
decorator==5.1.1
deeplabcut==2.3.3
dill==0.3.6
dlc2nwb==0.2
dlclibrary==0.0.2
dnspython==2.3.0
docopt==0.6.2
easydict==1.10
email-validator==1.3.1
entrypoints==0.4
etelemetry==0.3.0
executing==1.2.0
fasteners==0.18
filelock==3.10.6
filterpy==1.4.5
Flask==2.2.3
flatbuffers==23.3.3
fonttools==4.39.2
fqdn==1.5.1
fscacher==0.3.0
gast==0.4.0
google-auth==2.16.3
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
grpcio==1.51.3
h5py==3.8.0
hdmf==3.5.2
huggingface-hub==0.13.3
humanize==4.6.0
idna==3.4
imageio==2.27.0
imageio-ffmpeg==0.4.8
imgaug==0.4.0
importlib-metadata==6.1.0
interleave==0.2.1
ipykernel==6.22.0
ipython==8.11.0
isodate==0.6.1
isoduration==20.11.0
itsdangerous==2.1.2
jaraco.classes==3.2.3
jax==0.4.6
jedi==0.18.2
jeepney==0.8.0
Jinja2==3.1.2
jmespath==1.0.1
joblib==1.2.0
jsonpointer==2.3
jsonschema==4.17.3
jupyter_client==8.1.0
jupyter_core==5.3.0
keras==2.12.0
keyring==23.13.1
keyrings.alt==4.2.0
kiwisolver==1.4.4
lazy-ops==0.2.0
libclang==16.0.0
llvmlite==0.39.1
Markdown==3.4.3
MarkupSafe==2.1.2
matplotlib==3.7.1
matplotlib-inline==0.1.6
minio==7.1.14
more-itertools==9.1.0
msgpack==1.0.5
msgpack-numpy==0.4.8
natsort==8.3.1
ndx-pose==0.1.1
neo==0.12.0
nest-asyncio==1.5.6
networkx==2.6.3
neuroconv==0.2.4
num2words==0.5.12
numba==0.56.4
numcodecs==0.11.0
numexpr==2.8.4
numpy==1.23.5
nwbinspector==0.4.27
oauthlib==3.2.2
opencv-python-headless==4.7.0.72
opexebo==0.6.1
opt-einsum==3.3.0
otumat==0.3.1
packaging==23.0
pandas==1.5.3
parso==0.8.3
patsy==0.5.3
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.4.0
platformdirs==3.2.0
prompt-toolkit==3.0.38
protobuf==4.22.1
psutil==5.9.4
ptyprocess==0.7.0
pure-eval==0.2.2
py-cpuinfo==9.0.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pycryptodomex==3.17
pydantic==1.10.7
pydot==1.4.2
pyerfa==2.0.0.3
Pygments==2.14.0
PyMySQL==1.0.2
pynwb==2.3.1
pyout==0.7.2
pyparsing==3.0.9
pyrsistent==0.19.3
python-dateutil==2.8.2
pytz==2023.2
PyWavelets==1.4.1
PyYAML==6.0
pyzmq==25.0.2
quantities==0.14.1
requests==2.28.2
requests-oauthlib==1.3.1
rfc3339-validator==0.1.4
rfc3987==1.3.8
roiextractors==0.5.1
rsa==4.9
ruamel.yaml==0.17.21
ruamel.yaml.clib==0.2.7
s3transfer==0.6.0
scanimage-tiff-reader==1.4.1
scanreader @ git+https://github.com/kavli-ntnu/scanreader@96b8c6d2e62e0ad03a88582f8f8dfed94d3b476c
scikit-image==0.19.3
scikit-learn==1.2.2
scipy==1.10.1
SecretStorage==3.3.3
semantic-version==2.10.0
shapely==2.0.1
six==1.16.0
soupsieve==2.4
stack-data==0.6.2
statsmodels==0.13.5
tables==3.8.0
tabulate==0.9.0
tenacity==8.2.2
tensorboard==2.12.0
tensorboard-data-server==0.7.0
tensorboard-plugin-wit==1.8.1
tensorflow==2.12.0
tensorflow-estimator==2.12.0
tensorflow-io-gcs-filesystem==0.31.0
tensorpack==0.11
termcolor==2.2.0
tf-slim==1.1.0
threadpoolctl==3.1.0
tifffile==2023.3.21
torch==1.12.0
tornado==6.2
tqdm==4.65.0
traitlets==5.9.0
typing_extensions==4.5.0
uri-template==1.2.0
urllib3==1.26.15
watchdog==3.0.0
wcwidth==0.2.6
webcolors==1.13
Werkzeug==2.2.3
wrapt==1.14.1
zarr==2.14.2
zarr-checksum==0.2.8
zipp==3.15.0

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/catalystneuro/neuroconv/blob/main/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this bug was not already [reported](https://github.com/catalystneuro/neuroconv/issues)?

**Comment by simon-ball** [comment]
That addresses part of the bug, yielding a meaningful metadata dictionary:
```python
{'frameNumbers ': ' 1',
 'acquisitionNumbers ': ' 1',
 'frameNumberAcquisition ': ' 1',
 'frameTimestamps_sec ': ' 0.000000000',
 'acqTriggerTimestamps_sec ': ' -0.000088640',
 'nextFileMarkerTimestamps_sec ': ' -1.000000000',
 'endOfAcquisition ': ' 0',
 'endOfAcquisitionMode ': ' 0',
 'dcOverVoltage ': ' 0',
 'epoch ': ' [2023  2  6 10 51 13.374]',
 'auxTrigger0 ': ' []',
 'auxTrigger1 ': ' []',
 'auxTrigger2 ': ' []',
 'auxTrigger3 ': ' []',
 'I2CData ': ' {}'}
```

But, it misses the other part which is twofold:
* The metadata tag for which ScanImageExtractorInterface searches isn't present in the header of _any_ of the files we generate via ScanImage (tag `state.acq.frameRate`)
* None of the metadata that _is_ generated by ScanImage is extracted, because that's not stored in the field, in the tif, that's extracted (the field `scanimagetiffreader.ScanImageTiffReader(tif_file).description()` is extracted, but the SI metadata is present in `<...>.metadata()`)

Right now, I assume that either I'm spectacularly missing the point, or _every single_ researcher at Kavli working with Calcium imaging is consistently doing something weird with ScanImage. Because both of those two points above seem to be massive showstoppers that, presumably, wouldn't have made it through your testing without being noticed.

**Comment by bendichter** [comment]
@simon-ball we manage metadata extraction for about 38 different formats and base these programs on the files we receive. For some reason, the ScanImage files we received clearly do not match the files created by your group. This sort of thing happens all the time, so it's not very surprising. There was probably a software update somewhere along the way that changed this metadata.

As I mentioned previously, the ScanImage format is particularly difficult for us to manage and test, because the reader does not work on M1 macs. See an open issue for this [here](https://gitlab.com/vidriotech/scanimagetiffreader-python/-/issues/26). It would help if you could complain to ScanImage about this.

Is there any metadata in the ScanImage files you produce that would indicate a frame rate?

@CodyCBakerPhD do you think you could help with this? I am currently traveling and won't have access to my windows computer for another week and a half.

**Comment by CodyCBakerPhD** [comment]
Yes, I'd be happy to help with this issue, but I would need an example file to work off of

OR, if @simon-ball can confirm that the metadata in [these files](https://gin.g-node.org/CatalystNeuro/ophys_testing_data/src/main/imaging_datasets/ScanImage) labeled `scanimage_20220801_...` match their own style and access approach

**Comment by simon-ball** [comment]
Other priorities pushed this aside for a while, but I have not forgotten it.

Your repository includes 4 tif files like so:
1) scanimage_20220801_volume.tif
2) scanimage_20220801_multivolume.tif
3) scanimage_20220801_single.tif
4) sample_scanimage_version_3_8.tiff

I can confirm that 1-3 look superficially identical to my local examples, although I have not checked every single keyword. 

(In everything below, I have included the fix you provided as PR #386)

The current implementation of `ScanImageImagingInterface` already fails for test data 1-3 with the same error I encounter with the data generated by researchers here at KIN. 

``` python
ScanImageImagingInterface("scanimage_20220801_volume.tif")
>>> AssertionError: sampling frequency not found in image metadata, input the frequency using the argument `fallback_sampling_frequency`
```

Only test data 4 successfully reads the desired tag from its own exif data. If I understand the file name correctly, I assume that ScanImage _significantly_ changed how it stored its exif data between version 3.8 and the version 2020 that is currently in use here at Kavli, and whatever version was used to generate examples 1-3. 

Example 4 has metadata that looks like so:
``` python
from neuroconv.datainterfaces.ophys.scanimage.scanimageimaginginterface import extract_extra_metadata
extract_extra_metadata("sample_scanimage_version_3_8.tiff")
>>> {''state.configName': "'Behavior_2channel'",
 'state.software.version': '3.8',
 'state.software.minorRev': '0',
 'state.software.beta': '1',
 ....
}
```

Examples 1-3, and my own test data, look like so:
``` python
extract_extra_metadata("scanimage_20220801_multivolume.tif")
>>> {'frameNumbers': '8',
 'acquisitionNumbers': '1',
 'frameNumberAcquisition': '8',
 'frameTimestamps_sec': '0.233288390',
 'acqTriggerTimestamps_sec': '-1.000000000',
 'nextFileMarkerTimestamps_sec': '-1.000000000',
 'endOfAcquisition': '1',
 'endOfAcquisitionMode': '0',
 'dcOverVoltage': '0',
 'epoch': '[2022  8  8 16 56 7.329]',
 'auxTrigger0': '[]',
 'auxTrigger1': '[]',
 'auxTrigger2': '[]',
 'auxTrigger3': '[]',
 'I2CData': '{}',
}
```
In the (presumably) more modern version of ScanImage, metadata is arranged very differently. Firstly, and most importantly, it is no longer contained entirely within the `ScanImageTiffReader(file_path).description()` field. Most of the information is in `ScanImageTiffReader(file_path).metadata()` field.

Secondly, the exact tag names have changed. They're still broadly arranged in similar categories as for the older example (number 4), but as far as I can tell, both the path to specific tags has changed and, in some cases, the format of the value as well

If we look at one of your examples (and convert the string into a dictionary as is done already):

``` python
import ScanImageTiffReader
sr = ScanImageTiffReader.ScanImageTiffReader("scanimage_20220801_multivolume.tif")
{
          x.split("=")[0].strip(): x.split("=")[1].strip()
          for x in sr.metadata().replace("\n", "\r").split("\r")
          if "=" in x
      }
>>>'SI.LINE_FORMAT_VERSION': '1',
 'SI.PREMIUM': 'true',
 'SI.TIFF_FORMAT_VERSION': '4',
 'SI.VERSION_COMMIT': "'00ac849a74d5101cc34d41b1129ea49c1eb8b1d1'",
 'SI.VERSION_MAJOR': '2022',
 'SI.VERSION_MINOR': '0',
 'SI.VERSION_UPDATE': '0',
 'SI.acqState': "'grab'",
<400 lines ommitted for brevity>
}
```

---

So, solutions:

* Reading the metadata() field as well is a fairly trivial change, that's jsut processing oen more string from the exif data. This may not be the style you prefer, but it works. 

``` python
# neuroconv.datainterfaces.ophys.scanimage.scanimageimaginginterface
def extract_extra_metadata(file_path) -> dict:
    ScanImageTiffReader = get_package(
        package_name="ScanImageTiffReader", installation_instructions="pip install scanimage-tiff-reader"
    )
    src_file = ScanImageTiffReader.ScanImageTiffReader(str(file_path))
    extra_metadata = {}
    for metadata_string in (src_file.description(iframe=0), src_file.metadata()):
        metadata_dict = {
            x.split("=")[0].strip(): x.split("=")[1].strip()
            for x in metadata_string.replace("\n", "\r").split("\r")
            if "=" in x
        }
        extra_metadata = dict(**extra_metadata, **metadata_dict)
    return extra_metadata
```

* Reading the correct metadata tags is only as complex is identifying which are the relevant ones. This is one point I'm a bit lost on, as there are several tags that could be correct, and I don't have as much exposure to this particular aspect of data as some of the researchers. 
  * `state.acq.frameRate` - I _think_ this has been replaced by `"SI.hRoiManager.scanFrameRate"`, but I'm seeking confirmation, it might be `"SI.hRoiManager.scanVolumeRate"` instead
  * `"state.internal.triggerTimeString"` appears to be replaced by `epoch`, but annoyingly the format has changed too - it's a string representation of a list representation of a date time in typical big-endian datetime, instead of the ambigous, _maybe_ little-endian, _maybe_ USA-variant string. 
  * `'state.internal.triggerTimeString': "'10/9/2017 16:57:07.967'",` vs `'epoch': '[2022  8  8 16 56 7.329]',`

**Comment by simon-ball** [comment]
These points are addressed in #457, and all four example tifs (and my own example pass successfully, where "pass successfully" is defined as "can create the interface and invoke the `get_metadata()` method without raising an exception".
